### PR TITLE
Token Signing Algorithm Support

### PR DESCRIPTION
**Pull request by brouillette**
Proposal: update group membership to include the `alg` property , impacting
> `token_data` schema
> `IssuerAndClient-`and `IssuerAndSubject-` models, `SubjectClient` and `SubjectUser` types

### CODE REVIEW

**Comment by davidlougheed** [comment]
I don't think we should store algorithm in the database nor need to be able to run expressions against it. The list of valid JWT algorithms is provided by the OpenID config, and the correct algorithm is derived from there (albeit assuming the set of `id_token` algorithms works as a proxy for the access token JWTs).

**Comment by brouillette** [comment]
I think it's important that we can control for this.

As listed [here](https://www.loginradius.com/blog/engineering/jwt-signing-algorithms/) , for example, `HS256` is considered a valid signing algorithm, though it is considered much less secure than, say, `RS256`. From one bento instance to another, it'll surely come in handy to be able to filter out tokens an operator deems unfit from any particular provider at a more granular level, depending on the service being accessed.

**Comment by davidlougheed** [comment]
we can set up a way to forbid certain algorithms, that's a good point - but it doesn't make sense to put that at the grants level, since the grant applies to a subject, not a specific token (I am giving access to 'david'@'issuer.com'; the token algorithm doesn't factor into who I am)

one way to do what you're suggesting is to put a list of allowed algorithms in the config object and filter out any that don't match in `IdPManager` before decoding

File: bento_authorization_service/idp_manager.py [hunk 1]
```diff
@@ -11,16 +11,20 @@
...
     "IdPManager",
     "get_idp_manager",
     "IdPManagerDependency",
+    "check_token_signing_alg"
 ]
 
 
 class UninitializedIdPManagerError(Exception):
     pass
+class GeneralIdPManagerError(Exception):
```
Comment (davidlougheed) [inline bento_authorization_service/idp_manager.py]: code style: classes should be separated by exactly 2 newlines

Comment (davidlougheed) [reply bento_authorization_service/idp_manager.py]: also, rename this `IdPManagerError` and make it the base class for the previous class to nicely group related error types; then add a new subclass `IdPManagerBadAlgorithmError` and use that for token errors

File: bento_authorization_service/idp_manager.py [hunk 2]
```diff
@@ -90,7 +94,17 @@ async def decode(self, token: str) -> dict:
...
 
         # Assume we have the same set of signing algorithms for access tokens as ID tokens
-        return jwt.decode(token, sk, algorithms=self._oidc_well_known_data["id_token_signing_alg_values_supported"])
+        decoded_token = jwt.decode(token, sk, algorithms=self._oidc_well_known_data["id_token_signing_alg_values_supported"])
+
+        # Check if the algorithm used to sign the token is acceptable
+        check_token_signing_alg(decoded_token, config.permitted_token_algorithms)
+
+        return decoded_token
+
```
Comment (davidlougheed) [inline bento_authorization_service/idp_manager.py:180]: code style: 2 lines between function defs/classes etc at top level

File: bento_authorization_service/config.py [hunk 3]
```diff
@@ -27,6 +27,9 @@ class Config(BaseSettings):
     #    the instance-local IdP is supported.
     openid_well_known_url: str = "https://bentov2auth.local/realms/bentov2/.well-known/openid-configuration"
 
+    permitted_token_algorithms: frozenset = frozenset(["RS256", "ES256"])
```
Comment (davidlougheed) [inline bento_authorization_service/config.py]: add more supported algorithms by default - whatever is supported by keycloak (as alluded to below) minus whatever algorithms are insecure / what we don't want to support.

File: tests/test_evaluation.py [hunk 4]
```diff
@@ -65,6 +65,12 @@ def test_token_issuer_based_comparison():
         check_token_against_issuer_based_model_obj(sd.TEST_TOKEN, FakeIssBased(iss=sd.ISS))
 
 
+
+@pytest.mark.asyncio
+async def test_invalid_token_algo(db: Database, idp_manager: IdPManager, test_client: TestClient, db_cleanup):
+    with pytest.raises(GeneralIdPManagerError):
+        res = await evaluate(idp_manager, db, sd.make_fresh_david_no_alg_encoded(), sd.RESOURCE_PROJECT_1, frozenset({P_QUERY_DATA}))
+
```
Comment (davidlougheed) [inline tests/test_evaluation.py:81]: 2 newlines between functions

File: tests/test_evaluation.py [hunk 5]
```diff
@@ -65,6 +65,12 @@ def test_token_issuer_based_comparison():
         check_token_against_issuer_based_model_obj(sd.TEST_TOKEN, FakeIssBased(iss=sd.ISS))
 
 
+
+@pytest.mark.asyncio
+async def test_invalid_token_algo(db: Database, idp_manager: IdPManager, test_client: TestClient, db_cleanup):
+    with pytest.raises(GeneralIdPManagerError):
+        res = await evaluate(idp_manager, db, sd.make_fresh_david_no_alg_encoded(), sd.RESOURCE_PROJECT_1, frozenset({P_QUERY_DATA}))
```
Comment (davidlougheed) [inline tests/test_evaluation.py]: 120 char max line length (will add a linter!)

**Review by davidlougheed — CHANGES_REQUESTED** [review]

File: bento_authorization_service/config.py [hunk 6]
```diff
@@ -27,6 +27,11 @@ class Config(BaseSettings):
     #    the instance-local IdP is supported.
     openid_well_known_url: str = "https://bentov2auth.local/realms/bentov2/.well-known/openid-configuration"
 
+    permitted_token_algorithms: frozenset = frozenset(
+        ["RS256", "RSA-OAEP", "AES"]
```
Comment (davidlougheed) [inline bento_authorization_service/config.py]: RSA-OAEP is an encryption algorithm not a signing algorithm as far as I can tell. worth checking out https://auth.sd4h.ca/keycloak/realms/bentov2/.well-known/openid-configuration (specifically `id_token_signing_alg_values_supported` which should map pretty well to the access token too) to see which signing algorithms come with keycloak by default and then disabling any youve read are insecure.

**Review by davidlougheed — CHANGES_REQUESTED** [review]

File: tests/shared_data.py [hunk 7]
```diff
@@ -18,15 +18,48 @@
...
+
+TEST_DISABLED_TOKEN_SIGNING_ALGOS = [  # disable all supported algos except the strongest one
+    alg for alg in TEST_IDP_SUPPORTED_TOKEN_SIGNING_ALGOS if alg != TEST_TOKEN_SIGNING_ALG
+]
+
 TEST_TOKEN = {
     "iss": ISS,
     "sub": SUB,
     "aud": TEST_TOKEN_SECRET,
+    "alg": TEST_TOKEN_SIGNING_ALG,
```
Comment (davidlougheed) [inline tests/shared_data.py]: alg is not a JWT claim, it is in the header - so this shouldn't be included here / in token bodies.

**Review by davidlougheed — CHANGES_REQUESTED** [review]
one last thing and then good to merge

**Review by davidlougheed — APPROVED** [review]
lgtm

**Pull request merged**